### ISSUE TITLE
Warning message logs : Drop record, connection changed address.

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#993]
We deploy recently a version of Leshan using Cf 2.0.0-M14.

And I see some logs like this : 
```
--------------------------------
case 1 : 
2019-07-01 13:03:40.000
[ection-Handler-2] o.e.c.scandium.DTLSConnector Drop record Handshake (22), connection changed address /174.224.140.226:2070 => null!
2019-07-01 13:03:40.000
[ection-Handler-2] o.e.c.scandium.DTLSConnector Drop record Change Cipher Spec (20), connection changed address /174.224.140.226:2070 => null!
2019-07-01 13:03:40.000
[ection-Handler-2] o.e.c.scandium.DTLSConnector Drop record Handshake (22), connection changed address /174.224.140.226:2070 => null!
-------------------------------
case 2:
2019-07-01 13:28:40.000
[ection-Handler-4] o.e.c.scandium.DTLSConnector Drop record Alert (21), connection changed address /69.48.214.14:19042 => null!
2019-07-01 13:29:10.000
[ection-Handler-8] o.e.c.scandium.DTLSConnector Drop record Alert (21), connection changed address /69.48.214.14:19042 => null!
2019-07-01 13:29:59.000
[ction-Handler-11] o.e.c.scandium.DTLSConnector Drop record Alert (21), connection changed address /69.48.214.14:19042 => null!
...
```

I try to understand that but it's not so clear to me.
(We didn't use CID)

My understanding : 
When we received a packet/records we search corresponding connection in the store.

Then we process records in the serialexecutor context of this connection (so no concurrent access). The first thing we do in this execution context : checking that the connection is still associated to the record.

In this case, the peerAddress is null. This means that connection was removed from store between the moment we do the get the connection and the moment we process the record (very short time), right ? 
I'm not sure to understand how this could be possible.
1. maybe the connection is evicted ? This would be surprising because we have a EXPIRATION_THRESHOLD set to 10 minutes.
2. maybe the connection was removed because the handshake failed ?

For the case 1, I suspect the 3 messages are Client_key_exchange, Change_cipher_spec, finished. The psk is probably not good, so handshake failed, we remove connection from store, we didn't send alert because of psk oracle issue and so client retransmit it. The timing could maybe be explained by a slower access to pskstore.

For the case 2, this is more complicated there is 30s between each alert how this could happened ? I mean 3 race conditions in spaces of 60s :thinking: 

Did I missed something ?
(Maybe we should change the log by something as "connection removed from store", or something like this ?)

**Comment by boaks** [comment]
> In this case, the peerAddress is null. This means that connection was removed from store between the moment we do the get the connection and the moment we process the record (very short time), right ?

Yes, that is also my understanding, except the "very short time", which I would go to verify.

1. maybe the connection is evicted ? 
1. maybe the connection was removed because the handshake failed ?
1. maybe because the address is "reused" by a NAT for an other handshake?

I would first try to check the times. Should we add the time to the log? For 2.0.0-M16?

**Comment by boaks** [comment]
I'm not sure, but maybe issue #931, fixed with #932 is also a cause.
May be adding the time, when the connection lost it's address and the time to be executed provides more information about the case.

**Comment by sbernard31** [comment]
> except the "very short time", which I would go to verify.

My idea is : if the connection was removed since a long time, connectionStore.get would have return null.

> I would first try to check the times. Should we add the time to the log? For 2.0.0-M16?

Times are available in logs above ! or maybe you think about something else ?

> May be adding the time, when the connection lost it's address and the time to be executed provides more information about the case.

adding the time on the connection object or adding a log ?

> I'm not sure, but maybe issue #931, fixed with #932 is also a cause.

I would say this is not related, but I'm not so sure too. :sweat_smile: 

> Maybe we should change the log by something as "connection removed from store", or something like this ?

Does it make sense to you ?

**Comment by boaks** [comment]
Do you have logs, where you see the times of

```
get(address)
update(null)
log("drop")
```

Depending on the thread usage, between `get()` and `log("drop")` may be more time than expected.
And it would be also very interested, when the `update(null)` happens. therefore I thought to add these times explicit to the drop message log. Otherwise they must be collected from the logs, if the messages are in it.

**Comment by boaks** [comment]
> Does it make sense to you ?

That would be a other message for "null" then "other addresses". I would prefer, to find out what happens, and if PR #932 is related to it. After that, I would start to think about the log message.

**Comment by sbernard31** [comment]
> Do you have logs, where you see the times of

Nope, we do not activate DEBUG level (just WARN). Maybe we could give a try but I suspect this will generate a lot of logs !!

> therefore I thought to add these times explicit to the drop message log. 

It could makes sense.

This logs does not happen so often something like 3/6 by hours and most of the time this is more the alert pattern (case 2).

> if PR #932 is related to it

We will know that when we will deploy the next version of Leshan.

**Comment by boaks** [comment]
My feeling:
under stress retransmission are received faster than the serial jobs are executed.
CLIENT_HELLOs remove the previous connection (that's fixed with PR #932) and previous ALERTS may have removed the connection after the jobs for the duplicates are scheduled. So, maybe we introduce a "debug" mode, where we also store the record, if we have it, which removes the address from the connection, together with the time.

**Comment by boaks** [comment]
With 40s between the alerts, I'm not sure, if the executor is that busy.
Maybe a connection without address is still in the address-map.
The remove currently first updates the address to null and then removes it. 
If the update (during handshake?) throws an exception, the null-address connection sticks in the address-map.
I think, exchange that order of update and remove would be a additional good idea.
And a warning in the "Connection getBy()", if the resource is not longer referenced by the connection (as the address).

**Comment by sbernard31** [comment]
> Maybe a connection without address is still in the address-map.

This could be an explanation but I can not see how this could happen.

> If the update (during handshake?) throws an exception, the null-address connection sticks in the address-map.

If an exception was raised, it should be visible in log. we will activate INFO log on DTLSConnector, and I will let you know if I see something.
But reading the code I can not see how this could happen too.

> I think, exchange that order of update and remove would be a additional good idea.

I'm not sure to see which kind of exception could be raised between "setaddress(null)" and "remove from the store". Maybe the IllegalArgument one in updatePeerAddress ? (I don't really think this is possible as if `peerAddress` is null the exception must not be raised) (we should maybe not change the peeraddress if we raise the exception ?)

> And a warning in the "Connection getBy()", if the resource is not longer referenced by the connection (as the address).

I'm not sure to see which method you are talking about.

**Comment by boaks** [comment]
The connection store offers:

```
Connection find(final SessionId id);
Connection get(final InetSocketAddress peerAddress);
Connection get(final ConnectionId cid);
```

the intention is, to check, if the parameter is still in the returned connection.

**Comment by sbernard31** [comment]
> the intention is, to check, if the parameter is still in the returned connection.

Ok I get it. It makes sense.

---------------------------

We activated INFO log in DTLSConnector.
For now I didn't see any new logs when we face the Alert Drop (case 2).

For case 1 : handshake(22), change_cipher_spec(20), handshake(22), I faced only one case, no new related log too. 

But I face a new case which looks like (case 1) : change_cipher_spec(20), handshake(22).
```
2019-07-02 13:02:32.000
[ction-Handler-11] o.e.c.s.dtls.ServerHandshaker Cannot process handshake message from peer [/77.136.40.231:25683] due to [MAC validation failed] 
org.eclipse.californium.scandium.dtls.cipher.InvalidMacException: MAC validation failed at 
org.eclipse.californium.scandium.dtls.cipher.CCMBlockCipher.decrypt(CCMBlockCipher.java:355) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.dtls.Record.decryptAEAD(Record.java:697) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.dtls.Record.decryptFragment(Record.java:439) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.dtls.Record.getFragment(Record.java:977) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.dtls.Handshaker$InboundMessageBuffer.getNextMessage(Handshaker.java:333) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.dtls.Handshaker.processMessage(Handshaker.java:483) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector.processOngoingHandshakeMessage(DTLSConnector.java:1259) [av-lwm2m-bs.jar:na] at
org.eclipse.californium.scandium.DTLSConnector.processDecryptedHandshakeMessage(DTLSConnector.java:1245) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector.processHandshakeRecord(DTLSConnector.java:1209) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector.processRecord(DTLSConnector.java:930) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector.access$300(DTLSConnector.java:205) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector$7.run(DTLSConnector.java:848) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.elements.util.SerialExecutor$1.run(SerialExecutor.java:277) [av-lwm2m-bs.jar:na] at 
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_201] at 
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_201] at 
java.lang.Thread.run(Thread.java:748) [na:1.8.0_201]

2019-07-02 13:02:32.000
[ction-Handler-11] o.e.c.scandium.DTLSConnector Aborting handshake with peer [/77.136.40.231:25683]: Cannot process handshake message

2019-07-02 13:02:32.000
[ction-Handler-11] o.e.c.scandium.DTLSConnector Drop record Change Cipher Spec (20), connection changed address /77.136.40.231:25683 => null!

2019-07-02 13:02:32.000
[ction-Handler-11] o.e.c.scandium.DTLSConnector Drop record Handshake (22), connection changed address /77.136.40.231:25683 => null!
```
This looks like to what I imagine in https://github.com/eclipse/californium/issues/993#issue-462745234 for case 1.

**Comment by sbernard31** [comment]
Since yesterday, no INFO logs raised by DTLS connector which could be correlated to this drop record. (except the one above which occurred only 1 time)

The number of drop record is still constant (~3/6 per hour).

-------------------------------------------------
Not directly linked but just to let you know that since we migrate from leshan-1.0.0-M6(with cf-2.0.0-M9) to leshan-1.0.0-M11(with cf-2.0.0-M14)
Our CPU consumption go from ~6/7% to ~2/3% on our bootstrap server. :confetti_ball: 

-------------------------------------------------
Since we activate INFO log on DTLSconnector, I see 2 : 
```
[ection-Handler-6] o.e.c.scandium.DTLSConnector Unexpected error occurred while processing record from peer [/121.31.251.216:4575] java.lang.NegativeArraySizeException: null at
org.eclipse.californium.scandium.dtls.cipher.CCMBlockCipher.decrypt(CCMBlockCipher.java:316) ~[av-lwm2m-bs.jar:na] at
org.eclipse.californium.scandium.dtls.Record.decryptAEAD(Record.java:697) ~[av-lwm2m-bs.jar:na] at
org.eclipse.californium.scandium.dtls.Record.decryptFragment(Record.java:439) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.dtls.Record.getFragment(Record.java:977) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.dtls.Record.getFragment(Record.java:943) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector.processAlertRecord(DTLSConnector.java:1109) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector.processRecord(DTLSConnector.java:924) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector.access$300(DTLSConnector.java:205) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector$7.run(DTLSConnector.java:848) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.elements.util.SerialExecutor$1.run(SerialExecutor.java:277) [av-lwm2m-bs.jar:na] at 
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_201] at 
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_201] at 
java.lang.Thread.run(Thread.java:748) [na:1.8.0_201]
```
I feel this is just a "not well formed packet", but I'm not sure.
If you think this make sense to investigate this I can open a new issue.

**Comment by boaks** [comment]
> java.lang.NegativeArraySizeException

I think, add a "early drop" will do it better.

**Comment by boaks** [comment]
Do the clients use a "Alert Record" of epoch 1 without encryption?
I will add some logging to analyse that.

**Comment by sbernard31** [comment]
> Do the clients use a "Alert Record" of epoch 1 without encryption?

We don't know.

> I will add some logging to analyse that.

It could help.

**Comment by boaks** [comment]
`NegativeArraySizeException` 

that has the potential to terminate the connection in DTLSConnector.processRecord ...

**Comment by sbernard31** [comment]
A another log we never saw : 
```
[ection-Handler-2] .e.l.s.b.DefaultBootstrapHandler Error during bootstrap write of server instance LwM2mObjectInstance  on XXX org.eclipse.leshan.core.request.exception.SendFailedException: Unable to send request coap://50.58.14.5:7392/1/0 at
org.eclipse.leshan.core.californium.CoapAsyncRequestObserver.onSendError(CoapAsyncRequestObserver.java:102) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.core.coap.Message.setSendError(Message.java:855) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.core.coap.Request.setSendError(Request.java:810) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.core.network.CoapEndpoint$SendingCallback.onError(CoapEndpoint.java:964) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.elements.RawData.onError(RawData.java:278) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector$10.run(DTLSConnector.java:1714) [av-lwm2m-bs.jar:na] at 
org.eclipse.californium.elements.util.SerialExecutor$1.run(SerialExecutor.java:277) [av-lwm2m-bs.jar:na] at 
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_201] at 
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_201] at 
java.lang.Thread.run(Thread.java:748) [na:1.8.0_201]

Caused by: java.lang.NullPointerException: missing peer socket address, must not be null! at 
org.eclipse.californium.elements.AddressEndpointContext.<init>(AddressEndpointContext.java:87) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.elements.MapBasedEndpointContext.<init>(MapBasedEndpointContext.java:112) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.elements.MapBasedEndpointContext.<init>(MapBasedEndpointContext.java:81) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.elements.DtlsEndpointContext.<init>(DtlsEndpointContext.java:95) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.dtls.DTLSSession.getConnectionWriteContext(DTLSSession.java:419) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector.sendMessage(DTLSConnector.java:1834) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector.sendMessage(DTLSConnector.java:1825) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector.access$1100(DTLSConnector.java:205) ~[av-lwm2m-bs.jar:na] at 
org.eclipse.californium.scandium.DTLSConnector$10.run(DTLSConnector.java:1705) [av-lwm2m-bs.jar:na]
```
It seems this is still related to something like a call to `updatePeerAddress(null)`
Maybe because address was reused.
If this is a normal case we should maybe handle this.

I also see in DTLSConnector.sessionEstablished that there is a sendMessage which does not catch "RuntimeException". I think we should do that to report it to upper layer.

**Comment by boaks** [comment]
Hope, the most is included now. Anything of this issue missing for 2.0.0-M16? I'm looking forward to the next logging messages with the fixed record processing :-).

**Comment by boaks** [comment]
Please retest with 2.0.0-M16. 
I'm looking forward to your results.

**Comment by sbernard31** [comment]
Unfortunately, we will not be able to deploy this new version before end of August. :sob: 
(Too many vacation before this date :sweat_smile:  ...)

As soon as we do that I let you know what we found !

**Comment by boaks** [comment]
:-)

**Comment by sbernard31** [comment]
We have deployed cf-2.0.0-M16 since 5 days now and there is no more logs like this... 
Hard to know why... :thinking: 

I just saw for now  : 
1. two logs about :  "UdpMatcher new request could not be registered!"
2. one log about :   DTLSConnector (DTLS-0.0.0.0/0.0.0.0:5684) drops 166 bytes, IP(166.155.199.52:25683) != DTLS(166.155.199.52:25683,ID:5D6A48)"

I suppose this is about addresses reused. (e.g. because of NAT)
